### PR TITLE
Support get metadata and indexes API for REALTIME tables.

### PR DESCRIPTION
**Pull request by ramabme**
## Description
- Used segment.getDataSource(columnName) API which is implemented for both Mutable and Immutable segments.
- Removed the exposed private data API for ImmutableSegment.

Sample output for a REALTIME table segment (truncated json output)
```
{
  "airlineStats__8__0__20210716T1729Z": {
    "segmentName": "airlineStats__8__0__20210716T1729Z",
    "schemaName": "airlineStats",
    "crc": -9223372036854776000,
    "creationTimeMillis": 1626456582715,
    "creationTimeReadable": "2021-07-16T17:29:42:715 UTC",
    "timeGranularitySec": null,
    "startTimeMillis": null,
    "startTimeReadable": null,
    "endTimeMillis": null,
    "endTimeReadable": null,
    "segmentVersion": null,
    "creatorName": null,
    "custom": {},
    "indexes": {
      "FlightNum": {
        "bloom-filter": "NO",
        "dictionary": "YES",
        "forward-index": "YES",
        "inverted-index": "NO",
        "null-value-vector-reader": "NO",
        "range-index": "NO",
        "json-index": "NO"
      },
      "Origin": {
        "bloom-filter": "NO",
        "dictionary": "YES",
        "forward-index": "YES",
        "inverted-index": "NO",
        "null-value-vector-reader": "NO",
        "range-index": "NO",
        "json-index": "NO"
      },
      "Quarter": {
        "bloom-filter": "NO",
        "dictionary": "YES",
        "forward-index": "YES",
        "inverted-index": "NO",
        "null-value-vector-reader": "NO",
        "range-index": "NO",
        "json-index": "NO"
      },
    },
    "star-tree-index": null
  }
}
```
 
## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)

### CODE REVIEW

**Comment by mcvsubbu** [comment]
This seems to be fairly simple, I am wondering why realtime was omitted to begin with?

**Comment by mayankshriv** [comment]
> This seems to be fairly simple, I am wondering why realtime was omitted to begin with?

I wondered the same. I think looking at the original impl, the author ran into exceptions due to missing segmentVersion, and may have not pursued further to fix it. As long as this impl works, we are good. I also noticed that the original implementation exposed the internal map from metadata as opposed to use DataSource api's.

@ramabme Could you please paste a sample output in the PR description?

**Comment by mayankshriv** [comment]
@ramabme Could you please also check if your change continues to report the virtual columns, as the original implementation did.

**Comment by ramabme** [comment]
Cleaned up the functionality to return indexes only for the selected input columns.
Fixed the unit test to be consistent for the input number of columns.
And not returning index for the virtual columns.
(based on inputs from @Jackie-Jiang )

File: pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java [hunk 1]
```diff
@@ -76,78 +75,79 @@ public static String getSegmentMetadata(SegmentDataManager segmentDataManager, L
...
-      } else {
-        indexStatus.put(BLOOM_FILTER, INDEX_AVAILABLE);
-      }
-
-      if (Objects.isNull(columnIndexContainer.getDictionary())) {
-        indexStatus.put(DICTIONARY, INDEX_NOT_AVAILABLE);
-      } else {
-        indexStatus.put(DICTIONARY, INDEX_AVAILABLE);
-      }
+  private static void getSegmentColumnIndexes(String columnName, DataSource dataSource,
```
Comment (Jackie-Jiang) [inline pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java:105]: Suggest changing the signature to:
```
private static Map<String, String> getColumnIndexes(DataSource dataSource)
```

File: pinot-server/src/test/java/org/apache/pinot/server/api/TablesResourceTest.java [hunk 2]
```diff
@@ -264,18 +264,19 @@ public void testOfflineTableSegmentMetadata()
...
             .get(String.class));
     Assert.assertEquals(jsonResponse.get("columns").size(), 2);
-    Assert.assertEquals(jsonResponse.get("indexes").size(), 17);
+    Assert.assertEquals(jsonResponse.get("indexes").size(), 2);
     Assert.assertEquals(jsonResponse.get("star-tree-index").size(), 0);
 
     jsonResponse = JsonUtils.stringToJsonNode(
         (_webTarget.path(segmentMetadataPath).queryParam("columns", "*").request().get(String.class)));
     Assert.assertEquals(jsonResponse.get("columns").size(), segmentMetadata.getAllColumns().size());
+    Assert.assertEquals(jsonResponse.get("indexes").size(), 14);
```
Comment (Jackie-Jiang) [inline pinot-server/src/test/java/org/apache/pinot/server/api/TablesResourceTest.java:279]:
```suggestion
    Assert.assertEquals(jsonResponse.get("indexes").size(), segmentMetadata.getAllColumns().size());
```

**Review by Jackie-Jiang — APPROVED** [review]
LGTM with minor comments

**Comment by mayankshriv** [comment]
Curious, should we change the response to have the index size (instead of YES/NO), and also drop the ones which don't exist? That will reduce the overall response size. This doesn't need to be part of this PR.

**Comment by jadami10** [comment]
> and also drop the ones which don't exist? That will reduce the overall response size.

+1 to this. I'm looking at it in the UI now, and it's impossible to tell if a red X means "not reloaded" or "not indexed at all".

**Pull request merged**